File: core/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/player/playerUtils/PlaybackResumer.java
package com.pierfrancescosoffritti.androidyoutubeplayer.player.playerUtils;

import android.annotation.SuppressLint;
import androidx.annotation.NonNull;

import com.pierfrancescosoffritti.androidyoutubeplayer.player.PlayerConstants;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.YouTubePlayer;
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.AbstractYouTubePlayerListener;

/**
 * Class responsible for resuming the playback state in case of network problems.
 * eg: player is playing -> network goes out -> player stops -> network comes back -> player resumes playback automatically.
 */
public class PlaybackResumer extends AbstractYouTubePlayerListener {

    private boolean isPlaying = false;
    private PlayerConstants.PlayerError error = null;

    private String currentVideoId;
    private float currentSecond;

    public  void resume(YouTubePlayer youTubePlayer) {
        if(isPlaying && error == PlayerConstants.PlayerError.HTML_5_PLAYER)
            youTubePlayer.loadVideo(currentVideoId, currentSecond);
        else if(!isPlaying && error == PlayerConstants.PlayerError.HTML_5_PLAYER)
            youTubePlayer.cueVideo(currentVideoId, currentSecond);

        error = null;
    }

    @SuppressLint("SwitchIntDef")
    @Override
    public void onStateChange(@NonNull PlayerConstants.PlayerState state) {
        switch (state) {
            case ENDED:
                isPlaying = false;
                return;
            case PAUSED:
                isPlaying = false;
                return;
            case PLAYING:
                isPlaying = true;
                return;
            default:
                return;
        }
    }

    @Override
    public void onError(@NonNull PlayerConstants.PlayerError error) {
        if(error == PlayerConstants.PlayerError.HTML_5_PLAYER)
            this.error = error;
    }

    @Override
    public void onCurrentSecond(float second) {
        this.currentSecond = second;
    }

    @Override
    public void onVideoId(@NonNull String videoId) {
        this.currentVideoId = videoId;
    }
}

File: core/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/player/utils/PlaybackResumer.kt
package com.pierfrancescosoffritti.androidyoutubeplayer.player.utils

import android.annotation.SuppressLint

import com.pierfrancescosoffritti.androidyoutubeplayer.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.player.YouTubePlayer
import com.pierfrancescosoffritti.androidyoutubeplayer.player.listeners.AbstractYouTubePlayerListener
import com.pierfrancescosoffritti.androidyoutubeplayer.player.utils.YouTubePlayerUtils.loadOrCueVideo

/**
 * Class responsible for resuming the playback state in case of network problems.
 * eg: player is playing -> network goes out -> player stops -> network comes back -> player resumes playback automatically.
 */
internal class PlaybackResumer : AbstractYouTubePlayerListener() {

    private var isPlaying = false
    private var error: PlayerConstants.PlayerError? = null

    private var currentVideoId: String? = null
    private var currentSecond: Float = 0f

    fun resume(youTubePlayer: YouTubePlayer, canLoad: Boolean) {
        currentVideoId?.let { videoId ->
            if (isPlaying && error == PlayerConstants.PlayerError.HTML_5_PLAYER)
                youTubePlayer.loadOrCueVideo(canLoad, videoId, currentSecond)
            else if (!isPlaying && error == PlayerConstants.PlayerError.HTML_5_PLAYER)
                youTubePlayer.cueVideo(videoId, currentSecond)
        }

        error = null
    }

    override fun onStateChange(youTubePlayer: YouTubePlayer, state: PlayerConstants.PlayerState) {
        when (state) {
            PlayerConstants.PlayerState.ENDED -> {
                isPlaying = false
                return
            }
            PlayerConstants.PlayerState.PAUSED -> {
                isPlaying = false
                return
            }
            PlayerConstants.PlayerState.PLAYING -> {
                isPlaying = true
                return
            }
            else -> { }
        }
    }

    override fun onError(youTubePlayer: YouTubePlayer, error: PlayerConstants.PlayerError) {
        if (error == PlayerConstants.PlayerError.HTML_5_PLAYER)
            this.error = error
    }

    override fun onCurrentSecond(youTubePlayer: YouTubePlayer, second: Float) {
        currentSecond = second
    }

    override fun onVideoId(youTubePlayer: YouTubePlayer, videoId: String) {
        currentVideoId = videoId
    }
}
